format_test.go

```diff
@@ -393,3 +395,190 @@ func TestFormatJSONWithStack(t *testing.T) {
+		eris_test\.TestFormatJoinError:\S+:\d+
+		eris_test\.TestFormatJoinError:\S+:\d+
+	external
+2>	code\(unknown\) eris error
+		eris_test\.TestFormatJoinError:\S+:\d+`),
+			rootOutput: map[string]any{
+				"message": "join error",
+			},
+			wrapOutput: []map[string]any{
+				{
+					"message": "outer wrap",
+				},
+			},
+			externalsOutput: []map[string]any{
+				{
+					"external": "fmt error",
+				}, {
+					"external": "external",
+					"root":     map[string]any{},
+					"wrap":     []map[string]any{},
+				}, {
+					"root": map[string]any{},
+				},
+			},
+		},
+	}
+
+	for desc, tt := range tests {
+		t.Run(desc, func(t *testing.T) {
+			errStr := eris.ToCustomString(tt.input, eris.NewDefaultStringFormat(eris.FormatOptions{
+				WithTrace:    tt.withTrace,
+				WithExternal: tt.withExternal,
+			}))
+			if tt.stringOutput != "" && tt.stringOutput != errStr {
+				t.Errorf("%v: expected { %v } got { %v }", desc, tt.stringOutput, errStr)
+			}
+			if tt.regexOutput != nil && !tt.regexOutput.MatchString(errStr) {
+				t.Errorf("%v: expected match { %v } got { %v }", desc, tt.regexOutput.String(), errStr)
+			}
+
+			errJSON := eris.ToCustomJSON(tt.input, eris.NewDefaultJSONFormat(eris.FormatOptions{
+				WithTrace:    tt.withTrace,
+				WithExternal: tt.withExternal,
+			}))
+
+			// make sure messages are correct and stack elements exist (actual stack validation is in stack_test.go)
+			if rootMap, ok := errJSON["root"].(map[string]any); ok {
+				if _, exists := rootMap["message"]; !exists {
+					t.Fatalf("%v: expected a 'message' field in the output but didn't find one { %v }", desc, errJSON)
+				}
+				if rootMap["message"] != tt.rootOutput["message"] {
+					t.Errorf("%v: expected { %v } got { %v }", desc, tt.rootOutput["message"], rootMap["message"])
+				}
+				if _, exists := rootMap["stack"]; tt.withTrace && !exists {
+					t.Fatalf("%v: expected a 'stack' field in the output but didn't find one { %v }", desc, errJSON)
+				}
+			} else {
+				t.Errorf("%v: expected root error is malformed { %v }", desc, errJSON)
+			}
+
+			// make sure messages are correct and stack elements exist (actual stack validation is in stack_test.go)
+			if wrapMap, ok := errJSON["wrap"].([]map[string]any); ok {
+				if len(tt.wrapOutput) != len(wrapMap) {
+					t.Fatalf("%v: expected number of wrap layers { %v } doesn't match actual { %v }", desc, len(tt.wrapOutput), len(wrapMap))
+				}
+				for i := 0; i < len(wrapMap); i++ {
+					if _, exists := wrapMap[i]["message"]; !exists {
+						t.Fatalf("%v: expected a 'message' field in the output but didn't find one { %v }", desc, errJSON)
+					}
+					if wrapMap[i]["message"] != tt.wrapOutput[i]["message"] {
+						t.Errorf("%v: expected { %v } got { %v }", desc, tt.wrapOutput[i]["message"], wrapMap[i]["message"])
+					}
+					if _, exists := wrapMap[i]["stack"]; tt.withTrace && !exists {
+						t.Fatalf("%v: expected a 'stack' field in the output but didn't find one { %v }", desc, errJSON)
```

Comment:
nit: I suppose this can be `t.Errorf` as we can still check the rest of the wrapError in the loop.

Reply:
I copied the `wrapMap` section from other test cases. Leet's keep it 🤪